src/meds_torch/models/components/transformer_decoder.py

```diff
@@ -32,3 +298,286 @@ def forward(self, batch):
         batch[BACKBONE_TOKENS_KEY] = output
         batch[BACKBONE_EMBEDDINGS_KEY] = embeddings
         return batch
+
+    def generate(
+        self,
+        prompts: torch.Tensor,
+        mask: torch.Tensor | None = None,
+        get_next_token_time: Callable[[torch.Tensor, torch.Tensor], torch.Tensor] | None = None,
+        eos_token=None,
+        temperature=1.0,
+        filter_logits_fn: str | Callable = identity,
+        restrict_to_max_seq_len: bool = True,
+        filter_kwargs: dict = dict(),
+        cache_kv=True,
+        pad_value: int = 0,
+        **kwargs,
+    ) -> torch.Tensor:
+        """Generate tokens with either sequence length or time budget constraints.
+
+        Args:
+            prompts: Input token sequences [batch_size, seq_len]
+            budget: GenerationBudget specifying either max_seq_len or min_time_len
+            mask: Optional attention mask for prompts
+            temperature: Sampling temperature
+            filter_logits_fn: Name of logit filtering function ('top_k', 'top_p', etc.)
+            **kwargs: Additional arguments passed to the model
+
+        Returns:
+            torch.Tensor: Generated token sequences [batch_size, generated_len]
+        """
+        if mask is None:
+            prompt_lengths = torch.tensor([prompts.shape[1]] * prompts.shape[0])
+        else:
+            prompt_lengths = mask.sum(dim=-1)
+            right_pad_mask = torch.arange(mask.size(1), device=mask.device).unsqueeze(
+                0
+            ) < prompt_lengths.unsqueeze(1)
+            if not torch.equal(right_pad_mask, mask):
+                raise ValueError("Mask must correspond to right padding")
+
+        return generate(
+            self.model,
+            prompts=prompts,
+            budget=self.cfg.generation_budget,
+            get_next_token_time=get_next_token_time,
+            eos_token=eos_token,
+            temperature=temperature,
+            prompt_lens=prompt_lengths,
+            filter_logits_fn=filter_logits_fn,
+            restrict_to_max_seq_len=restrict_to_max_seq_len,
+            filter_kwargs=filter_kwargs,
+            cache_kv=cache_kv,
+            pad_value=pad_value,
+            **kwargs,
+        )
+
+
+@torch.no_grad()
+@eval_decorator
+def generate(
```

Comment:
This stuff seems very specific to the EIC case. should this be in a more specific file?